ambry-cloud/src/main/java/com/github/ambry/cloud/azure/AzureContainerCompactor.java

```diff
@@ -0,0 +1,132 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ */
+package com.github.ambry.cloud.azure;
+
+import com.azure.storage.blob.models.BlobErrorCode;
+import com.azure.storage.blob.models.BlobStorageException;
+import com.github.ambry.account.Container;
+import com.github.ambry.cloud.CloudRequestAgent;
+import com.github.ambry.cloud.CloudStorageException;
+import com.github.ambry.cloud.VcrMetrics;
+import com.github.ambry.config.CloudConfig;
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.nio.ByteBuffer;
+import java.util.Set;
+import java.util.stream.Collectors;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * Class that compacts containers in the Azure cloud by purging blobs of deleted containers from
+ * ABS and Cosmos.
+ */
+public class AzureContainerCompactor {
+  static final String CONTAINER_DELETION_CHECKPOINT_FILE = "container-deletion-checkpoint";
+  private static final Logger logger = LoggerFactory.getLogger(AzureContainerCompactor.class);
+
+  private final AzureBlobDataAccessor azureBlobDataAccessor;
+  private final CosmosDataAccessor cosmosDataAccessor;
+  private final CloudConfig cloudConfig;
+  private final VcrMetrics vcrMetrics;
+  private final AzureMetrics azureMetrics;
+  private final CloudRequestAgent requestAgent;
+
+  /**
+   * Constructor for {@link AzureContainerCompactor}.
+   * @param azureBlobDataAccessor {@link AzureBlobDataAccessor} object to access Azure Blob Store.
+   * @param cosmosDataAccessor {@link CosmosDataAccessor} object to access CosmosDb.
+   * @param cloudConfig {@link CloudConfig} object.
+   * @param vcrMetrics {@link VcrMetrics} object.
+   * @param azureMetrics {@link AzureMetrics} object.
+   */
+  public AzureContainerCompactor(AzureBlobDataAccessor azureBlobDataAccessor, CosmosDataAccessor cosmosDataAccessor,
+      CloudConfig cloudConfig, VcrMetrics vcrMetrics, AzureMetrics azureMetrics) {
+    this.azureBlobDataAccessor = azureBlobDataAccessor;
+    this.cosmosDataAccessor = cosmosDataAccessor;
+    this.cloudConfig = cloudConfig;
+    this.vcrMetrics = vcrMetrics;
+    this.azureMetrics = azureMetrics;
+    requestAgent = new CloudRequestAgent(cloudConfig, vcrMetrics);
+  }
+
+  /**
+   * Update newly deleted containers from {@code deletedContainers} to CosmosDb since last checkpoint.
+   * @param deletedContainers {@link Set} of deleted {@link Container}s.
+   * @throws CloudStorageException in case of any error.
+   */
+  public void updateDeletedContainers(Set<Container> deletedContainers) throws CloudStorageException {
```

Comment:
There's a situation that a container deletion has been canceled within the retention days so deleteTriggerTime will be back to 0, will you consider to sync that in the future implementation?